### PR TITLE
Fix #2832: Update onboarding flow.

### LINKED ISSUE #2832: Update onboarding flow
Related ticket 
https://github.com/brave/security/issues/199

### PR DESCRIPTION
**Pull request by iccub**
## Summary of Changes

Security review https://github.com/brave/security/issues/199

This pull request fixes #2832 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Screenshots:

Zrzut ekranu 2020-08-25 o 16 32 33

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Application/Delegates/AppDelegate.swift [hunk 1]
```diff
@@ -295,46 +295,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
                 Preferences.URP.referralLookupOutstanding.value = isFirstLaunch
             }
             
-            if Preferences.URP.referralLookupOutstanding.value == true {
```
Comment (iccub) [inline Client/Application/Delegates/AppDelegate.swift:298]: This chunk of code was not modified, it's moved to `handleReferralLookup`

File: Client/Application/Delegates/AppDelegate.swift [hunk 2]
```diff
@@ -344,6 +305,58 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
+        if initialOnboarding && UIPasteboard.general.hasStrings {
+            log.debug("Skipping URP call at app launch, this is handled in privacy consent onboarding screen")
+            return
+        }
+        
+        if Preferences.URP.referralLookupOutstanding.value == true {
+            var refCode: String?
+            
+            let savedRefCode: String? =
+                (!checkClipboard && ProcessInfo().operatingSystemVersion.majorVersion > 13) ?
```
Comment (kylehickinson) [inline Client/Application/Delegates/AppDelegate.swift]: Shouldn't this drop the `majorVersion` check now that we show privacy consent screen on all versions?

File: Client/Application/Delegates/AppDelegate.swift [hunk 3]
```diff
@@ -344,6 +305,58 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
+                    nil : UIPasteboard.general.string
```
Comment (kylehickinson) [inline Client/Application/Delegates/AppDelegate.swift]: Maybe also should leave a `FIXME` here reminding us to switch to iOS 14 APIs

File: Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift [hunk 4]
```diff
@@ -55,7 +56,20 @@ class OnboardingNavigationController: UINavigationController {
...
                     }
-                    return BraveAds.isCurrentLocaleSupported() ? [.searchEnginePicker, .shieldsInfo, .rewardsAgreement, .adsCountdown] : [.searchEnginePicker, .shieldsInfo, .rewardsAgreement]
+                    
+                    var newUserScreens: [Screens] = [.searchEnginePicker, .shieldsInfo, .rewardsAgreement]
+                    
+                    if BraveAds.isCurrentLocaleSupported() {
+                        newUserScreens.append(.adsCountdown)
+                    }
+                    
+                    // FIXME: Once iOS14 launches replace `hasStrings` with `UIPasteboard.DetectionPattern` api.
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift]: If we don't already have an issue made for fixing this and the other comment I made in iOS 14 then we might want to do that :)

File: Client/Application/Delegates/AppDelegate.swift [hunk 5]
```diff
@@ -344,6 +305,58 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
     }
+    
+    func handleReferralLookup(_ urp: UserReferralProgram, checkClipboard: Bool) {
+        let initialOnboarding =
+            Preferences.General.basicOnboardingProgress.value == OnboardingProgress.none.rawValue
+        
+        if initialOnboarding && UIPasteboard.general.hasStrings {
+            log.debug("Skipping URP call at app launch, this is handled in privacy consent onboarding screen")
+            return
+        }
```
Comment (kylehickinson) [inline Client/Application/Delegates/AppDelegate.swift:317]: Im not sure, why is this required? Cant we just only call it after onboarding? Anyone who has already onboarded should have already had URP checked, right?

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            $0.titleLabel?.adjustsFontSizeToFitWidth = true
+        }
+        
+        private let mainStackView = UIStackView().then {
+            $0.axis = .vertical
+            $0.distribution = .equalSpacing
+        }
+        
+        private let braveLogo = UIImageView(image: #imageLiteral(resourceName: "browser_lock_popup")).then { logo in
+            logo.contentMode = .scaleAspectFit
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift]: nit: Any reason for using named arg vs $0?

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        
+        private let copyFromClipboardLabel = UILabel().then {
+            $0.text = "Allow Brave to check my clipboard for a matching referral code"
+            $0.font = .systemFont(ofSize: 16, weight: .medium)
+            $0.numberOfLines = 0
+        }
+        
+        private let refProgramLabel = UILabel().then {
+            $0.text = Strings.OBPrivacyConsentDetail
+            $0.font = UIFont.systemFont(ofSize: 16, weight: .regular)
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift]: nit: Some of the font initializers you have using `UIFont.systemFont` and some just `.systemFont`, would prefer removing `UIFont` but as long as its consistent its fine

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            applyTheme(theme)
+            
+            mainStackView.addStackViewItems(
+                .view(.spacer(.vertical, amount: 1)),
+                .view(braveLogo),
+                .view(
+                    UIStackView(arrangedSubviews: [titleLabel, refProgramLabel]).then {
+                        $0.axis = .vertical
+                        $0.spacing = 24
+                }),
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift:62]: Could just do separate `.view` entries with a custom space after `titleLabel` instead of a nested stack view, i.e.

```
.view(titleLabel),
.customSpace(24),
.view(refProgramLabel),
```

Same for below

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Client/Application/Delegates/AppDelegate.swift (hunk 5 above)
Comment (iccub) [reply Client/Application/Delegates/AppDelegate.swift:317]: Not sure if I understand, but this check is for 2 cases:
- First launch, no ref code in clipboard, then we don't show the consent screen and URP logic is called from here
- First launch, ref code in clipboard, user kills the app while on the consent screen, on second launch we still wait with urp code until user goes past the privacy consent onboarding

File: Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift (hunk 4 above)
Comment (iccub) [reply Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift]: I will make a ticket, good idea

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift (hunk 8 above)
Comment (iccub) [reply Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift:62]: thanks, first time doing it your way with `addStackViewItems`

File: Client/Application/Delegates/AppDelegate.swift (hunk 5 above)
Comment (kylehickinson) [reply Client/Application/Delegates/AppDelegate.swift:317]: Ohhh gotcha, okay that makes sense thanks

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift (hunk 8 above)
Comment (iccub) [reply Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift:62]: actually it looks worse, because the `mainStackView` distribution is equal spacing, and I want to group labels and buttons to be close to each other

compare with the screenshot in PR description
Zrzut ekranu 2020-08-25 o 17 11 07

Comment (kylehickinson) [reply Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift:62]: Ahh I didn't notice that the main stack view had a different distribution. Carry on then :D

**Review by kylehickinson — APPROVED** [review]

File: Client/Application/Delegates/AppDelegate.swift [hunk 9]
```diff
@@ -295,46 +295,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
-                    }
-                    
-                    guard let url = offerUrl?.asURL else { return }
-                    self.browserViewController.openReferralLink(url: url)
-                }
-            } else {
-                urp.pingIfEnoughTimePassed()
-                browserViewController.backgroundDataSource.startFetching()
-            }
+            handleReferralLookup(urp, checkClipboard: false)
```
Comment (jumde) [inline Client/Application/Delegates/AppDelegate.swift:298]: @iccub - I'm not sure if I understand this call. If `checkClipboard` is false, `referral code` won't be accessed right?

File: Client/Application/Delegates/AppDelegate.swift [hunk 10]
```diff
@@ -344,6 +305,57 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
+            log.debug("Skipping URP call at app launch, this is handled in privacy consent onboarding screen")
+            return
+        }
+        
+        if Preferences.URP.referralLookupOutstanding.value == true {
+            var refCode: String?
+            
+            let savedRefCode = checkClipboard ? UIPasteboard.general.string : nil
+            
+            if Preferences.URP.referralCode.value == nil {
```
Comment (jumde) [inline Client/Application/Delegates/AppDelegate.swift:322]: @iccub - May be move this check before `UIPaseboard.general.string`?

File: Client/Application/Delegates/AppDelegate.swift [hunk 11]
```diff
@@ -344,6 +305,57 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
+                UrpLog.log("No ref code exists on launch, attempting clipboard retrieval")
+                refCode = UserReferralProgram.sanitize(input: savedRefCode)
+            }
+            
+            if refCode != nil {
```
Comment (jumde) [inline Client/Application/Delegates/AppDelegate.swift]: Move this in the `if Preferences.URP.referralCode.value == nil` block?

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            $0.contentMode = .scaleAspectFit
+        }
+        
+        private let titleLabel = CommonViews.primaryText(Strings.OBPrivacyConsentTitle).then {
+            $0.font = .systemFont(ofSize: 20, weight: .semibold)
+            $0.numberOfLines = 0
+        }
+        
+        private let copyFromClipboardLabel = UILabel().then {
+            $0.text = "Allow Brave to check my clipboard for a matching referral code"
```
Comment (jumde) [inline Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift]: @iccub - Shouldn't this be in BraveStrings.swift?

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentViewController.swift [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        contentView.yesConsentButton.addTarget(self, action: #selector(yesConsentTaapped), for: .touchUpInside)
+        contentView.noConsentButton.addTarget(self, action: #selector(noConsentTaapped), for: .touchUpInside)
+    }
+    
+    @objc func yesConsentTaapped() {
```
Comment (jumde) [inline Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentViewController.swift]: nit: Tapped

File: Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift [hunk 14]
```diff
@@ -55,7 +56,20 @@ class OnboardingNavigationController: UINavigationController {
...
-                    return BraveAds.isCurrentLocaleSupported() ? [.searchEnginePicker, .shieldsInfo, .rewardsAgreement, .adsCountdown] : [.searchEnginePicker, .shieldsInfo, .rewardsAgreement]
+                    
+                    var newUserScreens: [Screens] = [.searchEnginePicker, .shieldsInfo, .rewardsAgreement]
+                    
+                    if BraveAds.isCurrentLocaleSupported() {
+                        newUserScreens.append(.adsCountdown)
+                    }
+                    
+                    // FIXME: Update to iOS14 clipboard api once ready (#2838)
+                    if Preferences.URP.referralLookupOutstanding.value == true
```
Comment (jumde) [inline Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift]: @iccub - As I understand, the `privacyConsent` screen will be displayed multiple times if referral lookup fails but the code is read from the clipboard. I think it might make sense to assign `URP.referralCode.value` once the clipboard is accessed and then display the privacy consent screen only when `URP.referralCode.value == null`. Let me know if I'm missing something.

File: Client/Application/Delegates/AppDelegate.swift (hunk 9 above)
Comment (iccub) [reply Client/Application/Delegates/AppDelegate.swift:298]: Yes, if it is false we will try to do a IP-matching without accessing clipboard

File: Client/Application/Delegates/AppDelegate.swift (hunk 10 above)
Comment (iccub) [reply Client/Application/Delegates/AppDelegate.swift:324]: It has to be here, we sanitize the clipboard output after it has been read 
`UserReferralProgram.sanitize(input: savedRefCode)`

File: Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift (hunk 12 above)
Comment (iccub) [reply Client/Frontend/Browser/Onboarding/OnboardingPrivacyConsentView.swift]: this view is not used anymore, will delete

File: Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift (hunk 14 above)
Comment (iccub) [reply Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift]: Good idea

File: Client/Application/Delegates/AppDelegate.swift (hunk 10 above)
Comment (iccub) [reply Client/Application/Delegates/AppDelegate.swift:322]: I will move this, but in theory it should never happen

File: Client/Application/Delegates/AppDelegate.swift (hunk 9 above)
Comment (jumde) [reply Client/Application/Delegates/AppDelegate.swift:298]: That makes sense, but I'm confused about why this is needed if we already have this: https://github.com/brave/brave-ios/pull/2837/files#diff-d7437916ab14b41a34d7eb8e2a8d3afdR40

Comment (iccub) [reply Client/Application/Delegates/AppDelegate.swift:298]: @jumde  The privacy consent screen(`OnboardingPrivacyConsentViewController`) will not show if user has nothing in the clipboard, 
for this case we handle referral lookup here, and we force it to not access clipboard
so there's 3 ways to handle the referral now

1. No clipboard - do IP check here
2. Clipboard present - user consents - do clipboard referral lookup
3. Clipboard present - user denies - do IP referral lookup
